contracts/verifiers/UniversalVerifierMultiQuery.sol

```diff
@@ -0,0 +1,717 @@
+    modifier checkRequestGroupExistence(uint256 groupId, bool existence) {
+        if (existence) {
+            require(groupIdExists(groupId), "group id doesn't exist");
+        } else {
+            require(!groupIdExists(groupId), "group id already exists");
+        }
+        _;
+    }
+
+    /**
+     * @dev Modifier to check if the query exists
+     */
+    modifier checkQueryExistence(uint256 queryId, bool existence) {
+        if (existence) {
+            require(queryIdExists(queryId), "query id doesn't exist");
+        } else {
+            require(!queryIdExists(queryId), "query id already exists");
+        }
+        _;
+    }
+
+    /**
+     * @dev Modifier to check if the validator is whitelisted
+     */
+    modifier onlyWhitelistedValidator(IRequestValidator validator) {
+        require(isWhitelistedValidator(validator), "Validator is not whitelisted");
+        _;
+    }
+
+    /**
+     * @dev Initializes the contract
+     * @param state The state contract
+     * @param owner The owner of the contract
+     */
+    function initialize(IState state, address owner) public initializer {
+        __Ownable_init(owner);
+        _getUniversalVerifierMultiQueryStorage()._state = state;
+    }
+
+    function _getRequestType(uint256 requestId) internal pure returns (uint8 requestType) {
+        return uint8(requestId >> 248);
+    }
+
+    /**
+     * @dev Checks if a request ID exists
+     * @param requestId The ID of the request
+     * @return Whether the request ID exists
+     */
+    function requestIdExists(uint256 requestId) public view returns (bool) {
+        return
+            _getUniversalVerifierMultiQueryStorage()._requests[requestId].validator !=
+            IRequestValidator(address(0));
+    }
+
+    /**
+     * @dev Checks if a group ID exists
+     * @param groupId The ID of the group
+     * @return Whether the group ID exists
+     */
+    function groupIdExists(uint256 groupId) public view returns (bool) {
+        return _getUniversalVerifierMultiQueryStorage()._groupedRequests[groupId].length > 0;
+    }
+
+    /**
+     * @dev Checks if a query ID exists
+     * @param queryId The ID of the query
+     * @return Whether the query ID exists
+     */
+    function queryIdExists(uint256 queryId) public view returns (bool) {
+        return _getUniversalVerifierMultiQueryStorage()._queries[queryId].queryId == queryId;
+    }
+
+    function _getUniversalVerifierMultiQueryStorage()
+        private
+        pure
+        returns (UniversalVerifierMultiQueryStorage storage $)
+    {
+        assembly {
+            $.slot := UniversalVerifierMultiQueryStorageLocation
+        }
+    }
+
+    /**
+     * @dev Sets a request
+     * @param request The request data
+     */
+    function setRequest(
+        Request calldata request
+    )
+        public
+        checkRequestExistence(request.requestId, false)
+        checkRequestGroupExistence(request.validator.getGroupID(request.params), false)
```

Comment:
May be better to have `Request (struct)` and `bool` as parameters for `checkRequestGroupExistence`